doc/man/man1/gensky.1
.\" RCSid "$Id: gensky.1,v 1.6 2007/09/04 17:36:40 greg Exp $"
.TH GENSKY 1 4/24/98 RADIANCE
.SH NAME
gensky - generate a RADIANCE description of the sky
.SH SYNOPSIS
.B "gensky month day time"
[
.B options
]
.br
.B "gensky -ang altitude azimuth"
[
.B options
]
.br
.B "gensky -defaults"
.SH DESCRIPTION
.I Gensky
produces a RADIANCE scene description for the CIE standard
sky distribution at the given month, day and time.
By default, the time is interpreted as local standard
time on a 24-hour clock.
The time value may be given either as decimal hours, or using a
colon to separate hours and minutes.
If the time is immediately followed (no white space)
by a North American or European time zone designation,
then this determines the standard meridian, which may
be specified alternatively with the
.I \-m
option.
The following time zones are understood, with their corresponding
hour differences from Greenwich Mean Time:
.sp .5
.nf
Standard time:
YST   PST   MST   CST   EST   GMT
 9     8     7     6     5     0

CET   EET   AST   GST   IST   JST  NZST
\-1    \-2    \-3    \-4    \-5.5  \-9   \-12

Daylight savings time:
YDT   PDT   MDT   CDT   EDT   BST
 8     7     6     5     4     \-1

CEST  EEST  ADT   GDT   IDT   JDT  NZDT
 \-2    \-3   \-4    \-5   \-6.5   \-10  \-13
.fi
.PP
If the time is preceded by a plus sign ('+'), then it is interpreted
as local solar time instead.
It is very important to specify the correct latitude and longitude
(unless local solar time is given) using the
.I \-a
and
.I \-o
options to get the correct solar angles.
.PP
The second form gives the solar angles explicitly.
The altitude is measured in degrees above the horizon, and the
azimuth is measured in degrees west of South.
.PP
The third form prints the default option values.
.PP
The output sky distribution is given as a brightness function,
.I skyfunc.
Its value is in watts/steradian/meter2.
The x axis points east,
the y axis points north, and the z axis corresponds to the zenith.
The actual material and surface(s) used for the sky is left
up to the user.
For a hemispherical blue sky, the description might be:
.sp
.nf
!gensky 4 1 14

skyfunc glow skyglow
0
0
4 .99 .99 1.1 0

skyglow source sky
0
0
4 0 0 1 180
.fi
.sp
Often,
.I skyfunc
will actually be used to characterize the light coming in from
a window.
.PP
In addition to the specification of
a sky distribution function,
.I gensky
suggests an ambient value in a comment at the beginning of the
description to use with the
.I \-av
option of the RADIANCE rendering programs.
(See rvu(1) and rpict(1).)
This value is the cosine-weighted radiance of the sky in
watts/steradian/meter2.
.PP
.I Gensky
supports the following options.
.TP 10n
.BR \-s
Sunny sky without sun.
The sky distribution will correspond to a standard CIE clear day.
.TP
.BR \+s
Sunny sky with sun.
In addition to the sky distribution function, a source
description of the sun is generated.
.TP
.BR \-c
Cloudy sky.
The sky distribution will correspond to a standard CIE overcast day.
.TP
.BR \-i
Intermediate sky without sun.
The sky will correspond to a standard CIE intermediate day.
.TP
.BR \+i
Intermediate sky with sun.
In addition to the sky distribution, a (somewhat subdued) sun
is generated.
.TP
.BR \-u
Uniform cloudy sky.
The sky distribution will be completely uniform.
.TP
.BI -g \ rfl
Average ground reflectance is
.I rfl.
This value is used to compute
.I skyfunc
when Dz is negative.
Ground plane brightness is the same for
.I \-s
as for
.I \+s.
(Likewise for
.I \-i
and
.I \+i,
but see the
.I \-r
option below.)
.TP
.BI -b \ brt
The zenith brightness is
.I brt.
Zenith radiance (in watts/steradian/meter2) is normally computed
from the sun angle and sky turbidity (for sunny sky).
It can be given directly instead, using this option.
.TP
.BI -B \ irrad
Same as
.I \-b,
except zenith brightness is computed from the horizontal
diffuse irradiance (in watts/meter2).
.TP
.BI -r \ rad
The solar radiance is
.I rad.
Solar radiance (in watts/steradian/meter2) is normally computed from
the solar altitude.
This option may be used to override the default calculation.
If a value of zero is given, no sun description is produced, and the
contribution of direct solar to ground brightness is neglected.
.TP
.BI -R \ irrad
Same as
.I \-r,
except solar radiance is computed from the horizontal direct
irradiance (in watts/meter2).
.TP
.BI -t \ trb
The turbidity factor is
.I trb.
Greater turbidity factors
correspond to greater atmospheric scattering.
A turbidity factor of 1.0 indicates an ideal clear atmosphere (i.e.
a completely dark sky).
Values less than 1.0 are physically impossible.
.PP
The following options do not apply when the solar
altitude and azimuth are given explicitly.
.TP
.BI -a \ lat
The site latitude is
.I lat
degrees north.
(Use negative angle for south latitude.)
This is used in the calculation of sun angle.
.TP
.BI -o \ lon
The site longitude is
.I lon
degrees west.
(Use negative angle for east longitude.)
This is used in the calculation of solar time and sun angle.
Be sure to give the corresponding standard meridian also!
If solar time is given directly, then this option has no effect.
.TP
.BI -m \ mer
The site standard meridian is
.I mer
degrees west of Greenwich.
(Use negative angle for east.)
This is used in the calculation of solar time.
Be sure to give the correct longitude also!
If a time zone or solar time is given directly, then this option has no effect.
.SH EXAMPLE
To produce a sunny sky for July 4th at 2:30pm Eastern daylight time at a
site latitude of 42 degrees, 89 degrees west longitude:
.IP "" .2i
gensky 7 4 14:30EDT +s \-a 42 \-o 89
.PP
To produce a sunny sky distribution for a specific sun position but
without the sun description:
.IP "" .2i
gensky \-ang 23 \-40 \-s
.SH FILES
/usr/local/lib/ray/skybright.cal
.SH AUTHOR
Greg Ward
.SH "SEE ALSO"
rpict(1), rvu(1), xform(1)


src/gen/gensky.c
#ifndef lint
static const char	RCSid[] = "$Id: gensky.c,v 2.28 2020/07/25 19:18:01 greg Exp $";
#endif
/*
 *  gensky.c - program to generate sky functions.
 *		Our zenith is along the Z-axis, the X-axis
 *		points east, and the Y-axis points north.
 *		Radiance is in watts/steradian/sq. meter.
 *
 *     3/26/86
 */

#include  "rtio.h"
#include  <stdlib.h>
#include  <math.h>
#include  <ctype.h>
#include  "sun.h"
#include  "color.h"

#ifndef  PI
#define  PI		3.14159265358979323846
#endif

#define  DOT(v1,v2)	(v1[0]*v2[0]+v1[1]*v2[1]+v1[2]*v2[2])

#define  S_CLEAR	1
#define  S_OVER		2
#define  S_UNIF		3
#define  S_INTER	4

#define  overcast	((skytype==S_OVER)|(skytype==S_UNIF))

double  normsc();

#undef  toupper
#define  toupper(c)	((c) & ~0x20)	/* ASCII trick to convert case */

					/* European and North American zones */
struct {
	char	zname[8];	/* time zone name (all caps) */
	float	zmer;		/* standard meridian */
} tzone[] = {
	{"YST", 135}, {"YDT", 120},
	{"PST", 120}, {"PDT", 105},
	{"MST", 105}, {"MDT", 90},
	{"CST", 90}, {"CDT", 75},
	{"EST", 75}, {"EDT", 60},
	{"AST", 60}, {"ADT", 45},
	{"NST", 52.5}, {"NDT", 37.5},
	{"GMT", 0}, {"BST", -15},
	{"CET", -15}, {"CEST", -30},
	{"EET", -30}, {"EEST", -45},
	{"AST", -45}, {"ADT", -60},
	{"GST", -60}, {"GDT", -75},
	{"IST", -82.5}, {"IDT", -97.5},
	{"JST", -135}, {"NDT", -150},
	{"NZST", -180}, {"NZDT", -195},
	{"", 0}
};
					/* required values */
int  year = 0;					/* year (optional) */
int  month, day;				/* date */
double  hour;					/* time */
int  tsolar;					/* 0=standard, 1=solar */
double  altitude, azimuth;			/* or solar angles */
					/* default values */
int  skytype = S_CLEAR;				/* sky type */
int  dosun = 1;
double  zenithbr = 0.0;
int	u_zenith = 0;				/* -1=irradiance, 1=radiance */
double  turbidity = 2.45;
double  gprefl = 0.2;
					/* computed values */
double  sundir[3];
double  groundbr;
double  F2;
double  solarbr = 0.0;
int	u_solar = 0;				/* -1=irradiance, 1=radiance */

char  *progname;
char  errmsg[128];

void computesky(void);
void printsky(void);
void printdefaults(void);
void userror(char  *msg);
double normsc(void);
int cvthour(char  *hs);


int
main(
	int  argc,
	char  *argv[]
)
{
	int  got_meridian = 0;
	int  i;

	progname = argv[0];
	if (argc == 2 && !strcmp(argv[1], "-defaults")) {
		printdefaults();
		exit(0);
	}
	if (argc < 4)
		userror("arg count");
	if (!strcmp(argv[1], "-ang")) {
		altitude = atof(argv[2]) * (PI/180);
		azimuth = atof(argv[3]) * (PI/180);
		month = 0;
	} else {
		month = atoi(argv[1]);
		if (month < 1 || month > 12)
			userror("bad month");
		day = atoi(argv[2]);
		if (day < 1 || day > 31)
			userror("bad day");
		got_meridian = cvthour(argv[3]);
	}
	for (i = 4; i < argc; i++)
		if (argv[i][0] == '-' || argv[i][0] == '+')
			switch (argv[i][1]) {
			case 's':
				skytype = S_CLEAR;
				dosun = argv[i][0] == '+';
				break;
			case 'y':
				year = atoi(argv[++i]);
				break;
			case 'r':
			case 'R':
				u_solar = argv[i][1]=='R' ? -1 : 1;
				solarbr = atof(argv[++i]);
				break;
			case 'c':
				skytype = S_OVER;
				break;
			case 'u':
				skytype = S_UNIF;
				break;
			case 'i':
				skytype = S_INTER;
				dosun = argv[i][0] == '+';
				break;
			case 't':
				turbidity = atof(argv[++i]);
				break;
			case 'b':
			case 'B':
				u_zenith = argv[i][1]=='B' ? -1 : 1;
				zenithbr = atof(argv[++i]);
				break;
			case 'g':
				gprefl = atof(argv[++i]);
				break;
			case 'a':
				s_latitude = atof(argv[++i]) * (PI/180);
				break;
			case 'o':
				s_longitude = atof(argv[++i]) * (PI/180);
				break;
			case 'm':
				if (got_meridian) {
					++i;
					break;		/* time overrides */
				}
				s_meridian = atof(argv[++i]) * (PI/180);
				break;
			default:
				sprintf(errmsg, "unknown option: %s", argv[i]);
				userror(errmsg);
			}
		else
			userror("bad option");

	if (year && (year < 1950) | (year > 2050))
		fprintf(stderr,
			"%s: warning - year should be in range 1950-2050\n",
			progname);
	if (month && !tsolar && fabs(s_meridian-s_longitude) > 45*PI/180)
		fprintf(stderr,
	"%s: warning - %.1f hours btwn. standard meridian and longitude\n",
			progname, (s_longitude-s_meridian)*12/PI);

	fputs("# ", stdout);
	printargs(argc, argv, stdout);

	computesky();
	printsky();

	exit(0);
}


void
computesky(void)			/* compute sky parameters */
{
	double	normfactor;
					/* compute solar direction */
	if (month) {			/* from date and time */
		double  sd, st = hour;

		if (year) {			/* Michalsky algorithm? */
			double  mjd = mjdate(year, month, day, hour);
			if (tsolar)
				sd = msdec(mjd, NULL);
			else
				sd = msdec(mjd, &st);
		} else {
			int  jd = jdate(month, day);	/* Julian date */
			sd = sdec(jd);			/* solar declination */
			if (!tsolar)			/* get solar time? */
				st = hour + stadj(jd);
		}
		altitude = salt(sd, st);
		azimuth = sazi(sd, st);
		printf("# Local solar time: %.2f\n", st);
		printf("# Solar altitude and azimuth: %.1f %.1f\n",
				180./PI*altitude, 180./PI*azimuth);
	}
	if (!overcast && altitude > 87.*PI/180.) {
		fprintf(stderr,
"%s: warning - sun too close to zenith, reducing altitude to 87 degrees\n",
				progname);
		printf(
"# warning - sun too close to zenith, reducing altitude to 87 degrees\n");
		altitude = 87.*PI/180.;
	}
	sundir[0] = -sin(azimuth)*cos(altitude);
	sundir[1] = -cos(azimuth)*cos(altitude);
	sundir[2] = sin(altitude);

					/* Compute normalization factor */
	switch (skytype) {
	case S_UNIF:
		normfactor = 1.0;
		break;
	case S_OVER:
		normfactor = 0.777778;
		break;
	case S_CLEAR:
		F2 = 0.274*(0.91 + 10.0*exp(-3.0*(PI/2.0-altitude)) +
				0.45*sundir[2]*sundir[2]);
		normfactor = normsc()/F2/PI;
		break;
	case S_INTER:
		F2 = (2.739 + .9891*sin(.3119+2.6*altitude)) *
			exp(-(PI/2.0-altitude)*(.4441+1.48*altitude));
		normfactor = normsc()/F2/PI;
		break;
	}
					/* Compute zenith brightness */
	if (u_zenith == -1)
		zenithbr /= normfactor*PI;
	else if (u_zenith == 0) {
		if (overcast)
			zenithbr = 8.6*sundir[2] + .123;
		else
			zenithbr = (1.376*turbidity-1.81)*tan(altitude)+0.38;
		if (skytype == S_INTER)
			zenithbr = (zenithbr + 8.6*sundir[2] + .123)/2.0;
		if (zenithbr < 0.0)
			zenithbr = 0.0;
		else
			zenithbr *= 1000.0/SKYEFFICACY;
	}
					/* Compute horizontal radiance */
	groundbr = zenithbr*normfactor;
	printf("# Ground ambient level: %.1f\n", groundbr);
	if (!overcast && sundir[2] > 0.0 && (!u_solar || solarbr > 0.0)) {
		if (u_solar == -1)
			solarbr /= 6e-5*sundir[2];
		else if (u_solar == 0) {
			solarbr = 1.5e9/SUNEFFICACY *
			(1.147 - .147/(sundir[2]>.16?sundir[2]:.16));
			if (skytype == S_INTER)
				solarbr *= 0.15;	/* fudge factor! */
		}
		groundbr += 6e-5/PI*solarbr*sundir[2];
	} else
		dosun = 0;
	groundbr *= gprefl;
}


void
printsky(void)			/* print out sky */
{
	if (dosun) {
		printf("\nvoid light solar\n");
		printf("0\n0\n");
		printf("3 %.2e %.2e %.2e\n", solarbr, solarbr, solarbr);
		printf("\nsolar source sun\n");
		printf("0\n0\n");
		printf("4 %f %f %f 0.5\n", sundir[0], sundir[1], sundir[2]);
	}
	
	printf("\nvoid brightfunc skyfunc\n");
	printf("2 skybr skybright.cal\n");
	printf("0\n");
	if (overcast)
		printf("3 %d %.2e %.2e\n", skytype, zenithbr, groundbr);
	else
		printf("7 %d %.2e %.2e %.2e %f %f %f\n",
				skytype, zenithbr, groundbr, F2,
				sundir[0], sundir[1], sundir[2]);
}


void
printdefaults(void)			/* print default values */
{
	switch (skytype) {
	case S_OVER:
		printf("-c\t\t\t\t# Cloudy sky\n");
		break;
	case S_UNIF:
		printf("-u\t\t\t\t# Uniform cloudy sky\n");
		break;
	case S_INTER:
		if (dosun)
			printf("+i\t\t\t\t# Intermediate sky with sun\n");
		else
			printf("-i\t\t\t\t# Intermediate sky without sun\n");
		break;
	case S_CLEAR:
		if (dosun)
			printf("+s\t\t\t\t# Sunny sky with sun\n");
		else
			printf("-s\t\t\t\t# Sunny sky without sun\n");
		break;
	}
	printf("-g %f\t\t\t# Ground plane reflectance\n", gprefl);
	if (zenithbr > 0.0)
		printf("-b %f\t\t\t# Zenith radiance (watts/ster/m2\n", zenithbr);
	else
		printf("-t %f\t\t\t# Atmospheric turbidity\n", turbidity);
	printf("-a %f\t\t\t# Site latitude (degrees)\n", s_latitude*(180/PI));
	printf("-o %f\t\t\t# Site longitude (degrees)\n", s_longitude*(180/PI));
	printf("-m %f\t\t\t# Standard meridian (degrees)\n", s_meridian*(180/PI));
}


void
userror(			/* print usage error and quit */
	char  *msg
)
{
	if (msg != NULL)
		fprintf(stderr, "%s: Use error - %s\n", progname, msg);
	fprintf(stderr, "Usage: %s month day hour [options]\n", progname);
	fprintf(stderr, "   Or: %s -ang altitude azimuth [options]\n", progname);
	fprintf(stderr, "   Or: %s -defaults\n", progname);
	exit(1);
}


double
normsc(void)			/* compute normalization factor (E0*F2/L0) */
{
	static double  nfc[2][5] = {
				/* clear sky approx. */
		{2.766521, 0.547665, -0.369832, 0.009237, 0.059229},
				/* intermediate sky approx. */
		{3.5556, -2.7152, -1.3081, 1.0660, 0.60227},
	};
	double  *nf;
	double  x, nsc;
	int  i;
					/* polynomial approximation */
	nf = nfc[skytype==S_INTER];
	x = (altitude - PI/4.0)/(PI/4.0);
	nsc = nf[i=4];
	while (i--)
		nsc = nsc*x + nf[i];

	return(nsc);
}


int
cvthour(			/* convert hour string */
	char  *hs
)
{
	char  *cp = hs;
	int	i, j;

	if ( (tsolar = *cp == '+') ) cp++;		/* solar time? */
	while (isdigit(*cp)) cp++;
	if (*cp == ':')
		hour = atoi(hs) + atoi(++cp)/60.0;
	else {
		hour = atof(hs);
		if (*cp == '.') cp++;
	}
	while (isdigit(*cp)) cp++;
	if (!*cp)
		return(0);
	if (tsolar || !isalpha(*cp)) {
		fprintf(stderr, "%s: bad time format: %s\n", progname, hs);
		exit(1);
	}
	i = 0;
	do {
		for (j = 0; cp[j]; j++)
			if (toupper(cp[j]) != tzone[i].zname[j])
				break;
		if (!cp[j] && !tzone[i].zname[j]) {
			s_meridian = tzone[i].zmer * (PI/180);
			return(1);
		}
	} while (tzone[i++].zname[0]);

	fprintf(stderr, "%s: unknown time zone: %s\n", progname, cp);
	fprintf(stderr, "Known time zones:\n\t%s", tzone[0].zname);
	for (i = 1; tzone[i].zname[0]; i++)
		fprintf(stderr, " %s", tzone[i].zname);
	putc('\n', stderr);
	exit(1);
}
